timemachine/cpp/src/segmented_weighted_random_sampler.cu

```diff
@@ -0,0 +1,140 @@
+#include "device_buffer.hpp"
+#include "gpu_utils.cuh"
+#include "kernels/k_sampling.cuh"
+#include "kernels/kernel_utils.cuh"
+#include "math_utils.cuh"
+#include "segmented_weighted_random_sampler.hpp"
+
+namespace timemachine {
+
+template <typename RealType>
+SegmentedWeightedRandomSampler<RealType>::SegmentedWeightedRandomSampler(
+    const int max_vals_per_segment, const int num_segments, const int seed)
+    : max_vals_per_segment_(max_vals_per_segment), num_segments_(num_segments), temp_storage_bytes_(0),
+      d_gumbel_(round_up_even(max_vals_per_segment_ * num_segments_)), d_arg_max_(num_segments_), d_argmax_storage_(0) {
+
+    int *dummy_segments = nullptr;
+    gpuErrchk(cub::DeviceSegmentedReduce::ArgMax(
+        d_argmax_storage_.data,
+        temp_storage_bytes_,
+        d_gumbel_.data,
+        d_arg_max_.data,
+        num_segments_,
+        dummy_segments,
+        dummy_segments + 1));
+    d_argmax_storage_.realloc(temp_storage_bytes_);
+
+    curandErrchk(curandCreateGenerator(&cr_rng_, CURAND_RNG_PSEUDO_DEFAULT));
+    curandErrchk(curandSetPseudoRandomGeneratorSeed(cr_rng_, seed));
+};
+
+template <typename RealType> SegmentedWeightedRandomSampler<RealType>::~SegmentedWeightedRandomSampler() {
+    curandErrchk(curandDestroyGenerator(cr_rng_));
+};
+
+template <typename RealType>
+void SegmentedWeightedRandomSampler<RealType>::sample_device(
+    const int total_values,
+    const int num_segments,
+    const int *d_segment_offsets,        // [num_segments + 1]
+    const RealType *d_log_probabilities, // [total_values]
+    int *d_samples,                      // [num_segments]
+    cudaStream_t stream) {
+    // Perform sampling using the gumbel-max-trick like cupy
+    // https://github.com/cupy/cupy/blob/46f6ed6e78661fb1d31b41a072519f119f5a2385/cupy/random/_generator.py#L1115-L1121
+    curandErrchk(curandSetStream(cr_rng_, stream));
+    curandErrchk(templateCurandUniform(cr_rng_, d_gumbel_.data, d_gumbel_.length));
+
+    // Use the noise both as the noise and the buffer for the gumbel noise, does change the values in d_gumbel
+    this->sample_given_noise_device(
+        total_values,
+        num_segments,
+        d_segment_offsets,
+        d_log_probabilities,
+        d_gumbel_.data,
+        d_gumbel_.data,
```

Comment:
still confused about this

Reply:
@mcwitt in the old implementation, we directly overwrote the original `noise` values, but we require *all* input noise to be

1) generated upfront before a sequence of moves
2) be const (i.e. non-modifiable) 

we had a previous implementation that violated 2) (which is now addressed)

Reply:
> this function is only used for testing, in BD/TIBD we instead use sample_given_noise_device directly, which does not modify the input noise.

Currently only used in testing, but could be used in production where we don't care about being able to 'rewind' to previous noise.